src/main/java/com/puppycrawl/tools/checkstyle/site/XdocsTemplateParser.java

```diff
@@ -0,0 +1,200 @@
+                && !macroName.isEmpty()
+                && !isSecondParsing()) {
+            processMacroEnd(sink);
+            setIgnorableWhitespace(false);
+        }
+        else if (!tagName.equals(PARAM.toString())) {
+            sink.rawText(parser.getText());
+        }
+    }
+
+    /**
+     * Handle the opening tag of a macro. Gather the macro name and parameters.
+     *
+     * @param parser the xml parser.
+     * @throws MacroExecutionException if the macro name is not specified.
+     */
+    private void processMacroStart(XmlPullParser parser) throws MacroExecutionException {
+        macroName = parser.getAttributeValue(null, Attribute.NAME.toString());
+
+        if (macroName == null || macroName.isEmpty()) {
+            final String message = String.format(Locale.ROOT,
+                    "The '%s' attribute for the '%s' tag is required.",
+                    Attribute.NAME, MACRO_TAG);
+            throw new MacroExecutionException(message);
+        }
+    }
+
+    /**
+     * Handle the opening tag of a parameter. Gather the parameter name and value.
+     *
+     * @param parser the xml parser.
+     * @param sink the sink object.
+     * @throws MacroExecutionException if the parameter name or value is not specified.
+     */
+    private void processParamStart(XmlPullParser parser, Sink sink) throws MacroExecutionException {
+        if (macroName != null && !macroName.isEmpty()) {
+            final String paramName = parser
+                    .getAttributeValue(null, Attribute.NAME.toString());
+            final String paramValue = parser
+                    .getAttributeValue(null, Attribute.VALUE.toString());
+
+            if (paramName == null
+                    || paramValue == null
+                    || paramName.isEmpty()
+                    || paramValue.isEmpty()) {
+                final String message = String.format(Locale.ROOT,
+                        "'%s' and '%s' attributes for the '%s' tag are required"
+                                + "inside the '%s' tag.",
+                        Attribute.NAME, Attribute.VALUE, PARAM, MACRO_TAG);
+                throw new MacroExecutionException(message);
+            }
+
+            macroParameters.put(paramName, paramValue);
+        }
+        else {
+            sink.rawText(parser.getText());
+        }
+    }
+
+    /**
+     * Execute a macro. Creates a {@link MacroRequest} with the gathered
+     * {@link #macroName} and {@link #macroParameters} and executes the macro.
+     * Afterward, the macro fields are reinitialized.
+     *
+     * @param sink the sink object.
+     * @throws MacroExecutionException if a macro is not found.
+     */
+    private void processMacroEnd(Sink sink) throws MacroExecutionException {
+        final MacroRequest request = new MacroRequest(sourceContent,
+                new XdocsTemplateParser(), macroParameters,
+                new File(System.getProperty("user.dir")));
```

Comment:
Can we use constant here?

Reply:
Okay, I moved the `USER_DIR` constant to `CommonUtil`. Makes the most sense for me to be there.